### PR TITLE
Added menu component

### PR DESCRIPTION
**Pull request by nullhook**
Resolves: https://github.com/brave/leo/issues/368

TODO:
1.~Accept anchor element as a slot~
2. ~Make menu accessible~

### CODE REVIEW

File: src/components/menu/menu.stories.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.stories.svelte:47]: I don't think this component needs a label, or if it does, it should be part of the anchor

Comment (nullhook) [reply src/components/menu/menu.stories.svelte]: makes sense. I've removed this for now. However, we might need this in the future. ex:

image

Comment (petemill) [reply src/components/menu/menu.stories.svelte]: Do you mean the "Available language models" part? Would that just be a non-selectable item in the menu, provided in the default slot?

File: src/components/menu/menu.svelte
Comment (petemill) [inline src/components/menu/menu.svelte:115]: I guess we just need to make sure that the consumer doesn't provide their own button and only provides non-buttony content

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte]: nope, removed

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte]: removed

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte:30]: fixed

**Comment by nullhook** [comment]
@petemill added a "slots" story

Yes, that was the initial plan however the markup and quite different:

> The menu element's markup is different from that of the select's dropdown menu. The select menu has items with an accessibility state of 'selected' or 'not selected,' and the entire menu has a value. The standalone menu component created in this PR is a list of actionable items.


I think we can accept `leo-menu-item` if it's a standalone menu or `leo-option` if it's a select menu but then these needs to be differentiated roles for ex, `aria-selected` wont apply to menu items but leo-options, etc.

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte]: yes, removed

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [reply src/components/menu/menu.svelte:115]: I think even if they provide their own button this will be fine, as it has `all: unset`, so the styling should be fine :) Might be a bit weird from an accessibility perspective though

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:119]: So for slots, you can just provide the default value inside the slot:

```svelte
<slot name="anchor">
    <div>▶ Actions</div>
</slot>
```

I don't know if this default makes sense though - actions seems quite specific, and it won't be translated or anything.

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:36]: It worries me a bit that we're duplicating all this logic - the code is a bit finicky because it has to consider a bunch of different things, and I can pretty easily see a scenario where we fix a bug in one of these but not the other. It'd be nice if we could factor some of this out

**Review by fallaciousreasoning — COMMENTED** [review]
This looks pretty good, but I do worry about the amount of duplication between this and the Dropdown.

There's quite a bit of fiddly code for wrangling things across Svelte/Web Components/React and I worry about things getting out of sync between them.

I agree that semantically they're different, but from a code (and visual) standpoint, things seem quite similar. I wonder if it's worth trying to come up with a shared model?

@petemill wdyt? I'm also open to merging this in it's current form and trying to solve this in a separate PR, as @nullhook needs this

**Comment by fallaciousreasoning** [comment]
If we do decide to treat them as separate models, maybe we can do something more traditional, than the `<leo-menu-item>` components (i.e, with separate <Menu> and <MenuItem> components), so we don't need to do all the hackery with querySelector?

**Comment by petemill** [comment]
An example of one fork of the popup code getting enhancements and leaving the other fork in the dust is https://github.com/brave/leo/pull/384

Is it only the `aria-selected` that differs between the two? If so, that seems like 2 lines of code that could be controlled via a prop, e.g. `selectable`:

```svelte
  // When the options change, we should assign each one a tabindex - this let's
  // us handle changing the selection with the keyboard (tab or arrow keys).
  $: {
    for (const [option, index] of options.map((o, i) => [o, i] as const)) {
      option.setAttribute('tabindex', (index + 1).toString())
      if (selectable) {
        if (value === getValue(option)) option.setAttribute('aria-selected', '')
        else option.removeAttribute('aria-selected')
      } else {
         option.setAttribute('role', 'menu-item')
      }
    }
  }
```

**Comment by nullhook** [comment]
No, not just aria-selected but others too

**Standalone menu**

the anchor button:
```
aria-haspopup="true"
aria-controls="menu"
```

the menu:
```
role="menu"
```

the menu item:
```
role="menuitem"
```

**Select menu**
the label:
```

```
the option:

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte]: makes sense, removed

**Comment by nullhook** [comment]
I have rebased this PR on https://github.com/brave/leo/pull/384

After the rebase and update, the Menu component is designed to work with 'leo-menu-item' or 'leo-option'. It includes aria-related properties that are added to the respective elements. Additionally, it handles the current selected value, which will be useful for select components.

After #384 merges, I'll modify Dropdown component to use Menu internally.

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte:45]: We won't be duplicating it anymore. I'll use the Menu component after you merge the floating-ui PR.

File: src/components/button-menu/button-menu.stories.svelte
Comment (fallaciousreasoning) [reply src/components/button-menu/button-menu.stories.svelte:14]: Might be worth doing a bit more setup so it toggles on/off in the story

Comment (fallaciousreasoning) [reply src/components/button-menu/button-menu.stories.svelte:14]: Actually, it just looks like the build job hasn't been triggered - maybe it only runs for pulls against `main` which is a bit of a pain. Feel free to disregard.

Comment (fallaciousreasoning) [reply src/components/button-menu/button-menu.stories.svelte:14]: Okay, when I checked out locally it was similar, and toggling the `isOpen` prop on the storybook didn't seem to do anything

File: src/components/menu/menu.stories.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.stories.svelte:14]: I think let's just remove this button - not really needed for the`Menu` component

File: src/components/menu/menu.stories.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.stories.svelte:7]: Nit: remove

File: src/components/menu/menu.stories.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.stories.svelte:38]: Optional:
```suggestion
  leo-menu-item {
```

File: src/components/menu/menu.stories.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.stories.svelte:10]: Add `argTypes` prop and set the defaultValue for `isOpen` to `true` so the menu is open by default.

Comment (fallaciousreasoning) [reply src/components/menu/menu.stories.svelte:10]: See the `label` prop in `input.stories.svelte`

File: src/components/menu/menu.stories.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.stories.svelte:31]: Probably worth adding a `slots` story, for completeness

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:23]: Optional: I'm not 100% sure if you need the `= undefined` - you might be able to write:

```suggestion
  export let target?: HTMLElement
```

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:26]: Nit:
```suggestion
  const dispatch = createEventDispatcher()
```

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:49]: This probably needs to be `LEO-MENU-ITEM` now

Comment (fallaciousreasoning) [reply src/components/menu/menu.svelte:49]: Optionally this could be a prop, or we could come up with some cleverer way of doing this

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:61]: :+1:

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:72]: `leo-option` again

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:78]: :+1:

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:130]: Nit: add `autoUpdate` prop - this means if the user scrolls around the menu will do it's best to stay in view, while remaining anchored

File: src/components/menu/menu.svelte
Comment (fallaciousreasoning) [inline src/components/menu/menu.svelte:155]: :+1: I always forget this

**Review by fallaciousreasoning — APPROVED** [review]
This looks great @nullhook - I've left a bunch of comments, but none of them are super critical.

Before landing, could you do the following:
1. Make sure the Dropdown component in the `plain-html` example works (and ideally the react one too)
2. Make sure your new ButtonMenu component works in the `plain-html` example (and ideally the react one too)

You don't need to commit changes to the example projects, but you're welcome to push them if you want. It's more of a quick sanity check.

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte:130]: done

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte:10]: done

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte:38]: done

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte:14]: removed

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte:23]: i don't see this in the current codebase and I get a syntax error

File: src/components/menu/menu.stories.svelte
Comment (nullhook) [reply src/components/menu/menu.stories.svelte:31]: added

File: src/components/button-menu/button-menu.stories.svelte
Comment (nullhook) [reply src/components/button-menu/button-menu.stories.svelte:14]: ButtonMenu doesn't have isOpen exposed though?

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte:49]: Why LEO-MENU-ITEM? aria-selected is set for LEO-OPTION only, right?

Comment (fallaciousreasoning) [reply src/components/menu/menu.svelte:49]: Okay, sorry, I misunderstood how this was working - I assumed it was a holdover from the dropdown. Maybe instead of handling this via the type of the children, we could add a role prop to the `Menu` component, and do it based on that.

I worry that doing it based on the type of the children might be a bit non-obvious. WDYT @petemill?

File: src/components/button-menu/button-menu.stories.svelte
Comment (petemill) [reply src/components/button-menu/button-menu.stories.svelte:14]: Maybe we're talking about `Menu` here and not `ButtonMenu`. That has `isOpen` exposed on Storybook but it doesn't do anything. I think we don't need `Menu` on Storybook as it isn't a design system component in itself. It seems fine to have `ButtonMenu` and `Dropdown` exposed on Storybook and not confuse anyone with `Menu`.

Comment (nullhook) [reply src/components/button-menu/button-menu.stories.svelte:14]: That makes sense and I agree with Menu examples on storybook it basically ended up duplicating the logic of ButtonMenu. Anyways, I have removed Menu's storybook.

File: src/components/button-menu/button-menu.stories.svelte
Comment (petemill) [inline src/components/button-menu/button-menu.stories.svelte:30]: "dropdown" reference

File: src/components/menu/menu.svelte
Comment (petemill) [reply src/components/menu/menu.svelte:49]: It would be simpler for the users to simply only have leo-menu-item, with a role prop for the container to check for these conditional aria attrubutes.

One minor consideration would be menus whose popups behave more like lists, e.g. a "..." menu for choosing an AI model. It's basically a dropdown without the input. Should we allow that mechanism via choosing <leo-option> vs <leo-menu-item>? Or should it be via the container? What about a menu that has a mix of selectable options and menu-items? Will our devs be aware of this enough?

Happy to merge it like this if you think it's the way to go, though IMO let's specify in the Storybook when to use which. Also, are we missing `role="option"` for `leo-option` items?

File: src/components/menu/menu.svelte
Comment (petemill) [inline src/components/menu/menu.svelte:5]: Should we move `leo-option` definiton here too?

**Review by petemill — APPROVED** [review]
We can merge to https://github.com/brave/leo/pull/384 as that's blocked on https://github.com/brave/reviews/issues/1400 but required for this branch

File: src/components/button-menu/button-menu.stories.svelte
Comment (petemill) [reply src/components/button-menu/button-menu.stories.svelte:14]: I still see Menu in storybook 
image

Comment (petemill) [reply src/components/button-menu/button-menu.stories.svelte:14]: Looks like we didn't get a CI build after your latest push

Comment (petemill) [reply src/components/button-menu/button-menu.stories.svelte:14]: We can ignore that as it looks like it's due to https://github.com/brave/leo/pull/389

File: src/components/button-menu/button-menu.stories.svelte
Comment (nullhook) [reply src/components/button-menu/button-menu.stories.svelte]: removed

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte:5]: added

File: src/components/menu/menu.svelte
Comment (nullhook) [reply src/components/menu/menu.svelte:54]: I have added role=option for leo-option items. 

Let's see if the difference between 'leo-option' and 'leo-menu-item' is obvious to consumers, and I believe the first consumer for this would be AIChat. I have added a note regarding this in Storybook for both ButtonMenu and Dropdown.

**Pull request merged**
